Improved algebraic trait impls.

ryzhyk opened the pull request:
See commit messages.

Kixiron commented on src/algebra/finite_map.rs:
```diff
@@ -317,39 +317,102 @@ where
+                        e.get_mut().add_assign(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        };
```
```suggestion
                        }
```

Kixiron commented on src/algebra/finite_map.rs:
```diff
@@ -317,39 +317,102 @@ where
+                        e.get_mut().add_assign(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        };
```
```suggestion
                        }
```

Kixiron commented on src/algebra/finite_map.rs:
```diff
@@ -317,39 +317,102 @@ where
+                        e.get_mut().add_assign_by_ref(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        };
```
```suggestion
                        }
```

Kixiron commented on src/algebra/finite_map.rs:
```diff
@@ -317,39 +317,102 @@ where
+                        e.get_mut().add_assign_by_ref(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        };
```
```suggestion
                        }
```

Kixiron reviewed (commented):
I think the arms of the two sides could be factored out into functions where you just reverse the order of the arguments, that'd be nice in removing some of the duplicate code.
There's also some more if statements with semicolons around the code you changed that GitHub wouldn't let me suggest on if you feel like fixing those up.
I'm also not sure I like the `.support_size()` function name, it's not immediately apparent as to what it means

ryzhyk commented:
> I think the arms of the two sides could be factored out into functions where you just reverse the order of the arguments, that'd be nice in removing some of the duplicate code. There's also some more if statements with semicolons around the code you changed that GitHub wouldn't let me suggest on if you feel like fixing those up. I'm also not sure I like the `.support_size()` function name, it's not immediately apparent as to what it means

Thanks, fixed these issues and added some more optimizations.

mbudiu-vmw commented on src/algebra/finite_map.rs:
```diff
@@ -317,35 +315,109 @@ where
+        }
+
+        if self.support_size() > other.support_size() {
+            do_add(self.clone(), other)
```
both these do the same amount of work algorithmically. this optimization is justified only if clone is much faster than inserting all entries.

ryzhyk replied:
That's right, cloning the map should be more efficient than element-by-element insertion.
